spring-session-core/src/main/java/org/springframework/session/web/http/SessionRepositoryFilter.java

```diff
@@ -221,10 +221,11 @@ private void commitSession() {
 			}
 			else {
 				S session = wrappedSession.getSession();
+				String sessionId = session.getId();
+				boolean isRequestedSession = isRequestedSessionIdValid() && sessionId.equals(getRequestedSessionId());
 				clearRequestedSessionCache();
 				SessionRepositoryFilter.this.sessionRepository.save(session);
-				String sessionId = session.getId();
```

Comment:
Is the sessionId allowed to change on save? In that case, the comparison should still be done after save, but requestedSessionId has to be retrieved before the clearRequestedSessionCache()